### PR TITLE
Composable index templates for ES 7.8

### PR DESCRIPTION
**Pull request by ccharnkij**
# Description

For #3138, this PR lets Zipkin uses composable index templates for ES >= 7.8. Also, a new env variable is added for index template priority.

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# How Has This Been Tested?

Unit tests are added for the features. The unit tests passed when ran locally. Also, I have done E2E tests using both ES >=7.8 and ES < 7.8 using the brave sample project.

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/VersionSpecificTemplates.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/VersionSpecificTemplates.java:98]: beginTemplate

File: zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/VersionSpecificTemplates.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/VersionSpecificTemplates.java]: squash this and below into endTemplate as it is distracting to add template priority here when sometimes there's no template anyway.

File: zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/ElasticsearchStorageTest.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/ElasticsearchStorageTest.java:245]: sorry I don't understand what is special checked here, as unless blind, I would expect the same to pass before 7.8?

In any case, if we are verifying these are legacy, we should mention that in the test name or description, as currently the name doesn't seem to suggest what we are looking at (unless I miss something)

File: zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/VersionSpecificTemplatesTest.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/VersionSpecificTemplatesTest.java:139]: nice. I suppose at some point we should switch this assert to json path, but works for me.

**Review by codefromthecrypt — COMMENTED** [review]
This looks very good!

I would add something in the ITElasticsearch7 similar to ITEnsureSchema in cassandra, that sets up something that this priority would ensure doesn't mess up. Ex start ES docker and add the template prior to zipkin doing so. then make sure the things that failed before doesn't.

This way, we can automatically ensure something like @benqs comment doesn't hit us on 7.9 https://github.com/openzipkin/zipkin-support/issues/23#issuecomment-654066659

Also add a test for version 7.9 is same behavior as sometimes eyes cross on expressions. This happened to us in cassandra before.

File: zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/ElasticsearchStorageTest.java
Comment (ccharnkij) [reply zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/ElasticsearchStorageTest.java:245]: Hi,

This test is for checking that Zipkin uses resource path of "/_index_template/....", as opposed to "/_template/...", for ES >= 7.8 when creating index templates. I can remove it if it's not needed. Or I can change test name (or add comment) for it to be more descriptive.

Comment (codefromthecrypt) [reply zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/ElasticsearchStorageTest.java:245]: yeah you can clarify with comment why it is important _index_template as I believe in some cases it worked before with _template. composable is a bit jargon to a passer-by.

**Comment by ccharnkij** [comment]
@adriancole Please review again. I made begin/endTemplate and added commented on tests. I also added to ITElasticsearch7  as you suggested, I created a 'catch-all' template with pattern '*' and priority 0 prior to the tests, and the tests should proceed as normal.

I know that there are license fail checks, but since I didn't add any new file, I wasn't sure about updating the license header.

Also, I wasn't sure what you mean about tests for version 7.9. The code/tests that I added should apply to version 7.8 and above. Unless I'm missing something, please let me know.

**Comment by codefromthecrypt** [comment]
FYI command to update license is here https://github.com/openzipkin/zipkin/blob/master/.github/CONTRIBUTING.md

File: zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/ElasticsearchStorage.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/ElasticsearchStorage.java]: this comment doesn't belong. probably you want to say `// highest priority`

File: zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/ElasticsearchStorageTest.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/ElasticsearchStorageTest.java:235]: for example, eaisiest way is to change this to 7.9

File: zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITElasticsearchStorageV7.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITElasticsearchStorageV7.java]: if using square formatter it will expand wildcard imports

File: zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ElasticsearchStorageExtension.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ElasticsearchStorageExtension.java]: maybe revert this as it effects all tests.. that or comment why.

File: zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITElasticsearchStorageV7.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITElasticsearchStorageV7.java]: should be a separate test instead of a before for everything.

**Review by codefromthecrypt — COMMENTED** [review]
thanks for the revision. Ideally the composite test is not implicit, rather a separate test. if the version is <7.8 you can use https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assumptions.html#assumeTrue(java.util.function.BooleanSupplier)

**Comment by ccharnkij** [comment]
@adriancole 3rd time's a charm?

File: zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITElasticsearchStorage.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITElasticsearchStorage.java]: I think you can move this to the ITEs7 class?

File: zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java]: with this pinned to the ES 7 IT, you don't need to check version as our image is already this I think right? In this case you can do this in the test method itself

File: zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java]: this should be  a "finally" of the catchAllTemplateTest as it would only need to be tested once anyway

File: zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java]: verify the catch-all template works? you might want to make something more interesting so that you can use ES HTTP query on it. ex this one https://gist.github.com/adriancole/1af1259102e7a2da1b3c9103565165d7

**Review by codefromthecrypt — COMMENTED** [review]
Thanks I think you did the hard work. Main thing is to isolate the catch-all to a single test and also re-scope it so that it would only ever invoke on ES 7.8+

**Comment by ccharnkij** [comment]
@adriancole like that?

File: zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITElasticsearchStorageV7.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITElasticsearchStorageV7.java]: as other tests we add to ITEnsureIndexTemplate won't necessarily effect templatePriority, I would move that setting into the test itself. Then, you can use the same setup as the other tests like this:

```java
    @Override protected ElasticsearchStorage.Builder newStorageBuilder(TestInfo testInfo) {
      return backend().computeStorageBuilder().index(index(testInfo));
    }

    @AfterEach public void clear() throws IOException {
      storage.clear();
    }
```

File: zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java]: nit protected not needed in the same package

File: zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java]: if this is the catch all change you want to make, it would be useful to test that this applies. The test could pass for a subtle reason, then in ES 7.10 it might break and we wouldn't know what it was trying to do either.

For this reason, making a template that sets default which I think this does, maybe isn't a good choice. 
Instead try something with semantic value and also relevant in tracing? ex
```
  "mappings": {
    "properties": {
      "tags.http.status_code": {
        "norms": false,
        "type": "keyword"
      }
    }
  }
```

In this case, your test can switch from making a fake tag "queryTest"->"ok"  to a tag very commonly requested  like "http.status_code" "404"

Then, you can verify your catch-all worked by using HTTP to do the following (except in java):

```
curl -s 'localhost:9200/zipkin-span-*/_search?q=tags.http.status_code:404'
```

Ex what you are doing is literally verifying automatically this https://gist.github.com/adriancole/1af1259102e7a2da1b3c9103565165d7

**Review by codefromthecrypt — COMMENTED** [review]
Almost there! let me know if you are out of steam, but otherwise it would be great if you can finish this. Thanks again

File: zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java
Comment (ccharnkij) [reply zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java]: I don't think the suggested secondary template works anymore, given that ES no longer allows multiple matching index templates of the same priority. ES won't merge templates based on order anymore, and will only select 1 with the highest priority. Multiple index template with matching index patterns and priorities will result in error on PUT of the second index template. So in this case, it's either Zipkin's template or catch-all template, and I can't sneak in tags to zipkin-span-* like that.

I do see what you mean, though. And if the objective is to test the catch-all, I think I have to use a new index that doesn't match Zipkin's template at all. Thought?

Comment (codefromthecrypt) [reply zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java]: interesting.. thanks for the info! yeah a comment should say this is the case now ( I remember hints of this in the past from @xeraa ). I do think the default template should meddle with ours in a way that would be obvious then. For example, you could change something we don't set? then read back that it is unset (as opposed to inherited/merged) would that do it? 

meanwhile, using http in your test is still probably better as it is more realistic data, but yeah I can now expect people who setup multiple templates to have to do an offline merge of them instead of my gist now!

Comment (xeraa) [reply zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java]: 1. There are the [legacy / v1](https://www.elastic.co/guide/en/elasticsearch/reference/7.9/indices-templates-v1.html) templates that can be merged, but I would be careful (multiple templates with the same order will result in a non-deterministic merging order, merging valid templates might lead to an invalid mapping). Right now they are deprecated and we'll see when we'll remove them — you have at least until 8.0.
2. The new [component based approach](https://www.elastic.co/guide/en/elasticsearch/reference/7.9/index-templates.html) still allows you to combine multiple pieces, but you don't merge them (to avoid the two problems mentioned above). This feature was only added in 7.8, so I'm not sure what will be the sanest approach for you — support it in 7.x already or switch over for any (upcoming) 8.x cluster.

Comment (codefromthecrypt) [reply zipkin-storage/elasticsearch/src/test/java/zipkin2/elasticsearch/integration/ITEnsureIndexTemplate.java]: ok so legacy/v1 would "for some time" retain the current behavior. I suppose in this case this means people in-place upgrading continue to work as this logic only applies when the templates don't already exist. cc also @basvanbeek for perspective on your schema tool. This chat will likely outgrow this PR :P

@ccharnkij for you and to not cage you to this decision tree, I would make a comment to the effect that @xeraa made in point 1. in your unit test and possibly also revise the javadoc/README if doesn't already say this. Basically, this relies on that we are not using legacy/v1 templates, so it is a win, not a merge.

hence we expect them to not merge. So, as mentioned earlier to show that actually happened (sanity check), have your template override something we don't and then magically notice that after the storage component initializes, whatever that setting is, is now ignored.

**Comment by ccharnkij** [comment]
@adriancole how about this one? 

Regarding the legacy vs new template though, perhaps I could help more with it. I could allow both configuration based on environment variable, maybe ES_TEMPLATE_TYPE=legacy or new (or something like that). For ES >= 7.8, people can choose if they still want to use the legacy one or switch over to the one. Thought?

**Comment by codefromthecrypt** [comment]
Thanks for the volunteer to move forward. I'm a fan of least config possible.

The template priority is a new config. Maybe we can use this in such a way that if you set it, you accept using "new templates" failing that defaults to legacy ones. If this is logical, it clears up the mixed state of 7.x and at any rate 8.0 won't support legacy ones. wdyt?

**Comment by ccharnkij** [comment]
That works too. I just have to remove the implicit default priority=0 when not set. People that want to use new template just have to set priority all the time even if they want priority=0, until legacy template are completely removed.

**Comment by codefromthecrypt** [comment]
yep sgtm!

**Comment by ccharnkij** [comment]
@adriancole how about this one?

File: zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/ElasticsearchStorage.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/ElasticsearchStorage.java]: probably using Zipkin's Nullable for Integer is better, just to prevent a little brain break

File: zipkin-server/src/main/resources/zipkin-server-shared.yml
Comment (codefromthecrypt) [inline zipkin-server/src/main/resources/zipkin-server-shared.yml]: to match your docs..

```suggestion
      priority: ${ES_TEMPLATE_PRIORITY:}
```

File: zipkin-server/src/main/java/zipkin2/server/internal/elasticsearch/ZipkinElasticsearchStorageProperties.java
Comment (codefromthecrypt) [inline zipkin-server/src/main/java/zipkin2/server/internal/elasticsearch/ZipkinElasticsearchStorageProperties.java]: make Integer

File: zipkin-server/src/main/java/zipkin2/server/internal/elasticsearch/ZipkinElasticsearchStorageProperties.java
Comment (codefromthecrypt) [inline zipkin-server/src/main/java/zipkin2/server/internal/elasticsearch/ZipkinElasticsearchStorageProperties.java]: If this is a string to coerce empty to null, make a comment (also coerce empty to null :P) but still use an integer in the builder. I'm not actually sure what Spring does to an integer field set to ""

**Review by codefromthecrypt — COMMENTED** [review]
Looks good. one minor thing and I think we're done

File: zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/ElasticsearchStorage.java
Comment (codefromthecrypt) [inline zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/ElasticsearchStorage.java]:
```suggestion
    public abstract Builder templatePriority(@Nullable Integer priority);
```

File: zipkin-server/src/main/resources/zipkin-server-shared.yml
Comment (codefromthecrypt) [inline zipkin-server/src/main/resources/zipkin-server-shared.yml]: sorry I didn't notice this earlier: rename to template-priority

add a test in ZipkinElasticsearchStorageConfigurationTest to ensure that `zipkin.storage.elasticsearch.template-priority:` coerces to null not zero

File: zipkin-server/src/main/java/zipkin2/server/internal/elasticsearch/ZipkinElasticsearchStorageProperties.java
Comment (codefromthecrypt) [inline zipkin-server/src/main/java/zipkin2/server/internal/elasticsearch/ZipkinElasticsearchStorageProperties.java]: rename to templatePriority

File: zipkin-server/src/main/java/zipkin2/server/internal/elasticsearch/ZipkinElasticsearchStorageProperties.java
Comment (codefromthecrypt) [inline zipkin-server/src/main/java/zipkin2/server/internal/elasticsearch/ZipkinElasticsearchStorageProperties.java]: rename to template-priority

**Review by codefromthecrypt — COMMENTED** [review]
sorry I didn't notice the naming thing earlier

File: zipkin-server/src/main/resources/zipkin-server-shared.yml
Comment (ccharnkij) [reply zipkin-server/src/main/resources/zipkin-server-shared.yml:167]: Would you prefer ES_TEMPLATE_PRIORITY= <--- (aka null)  to default to 0, instead of using legacy template? With the type set to nullable Integer, if the ES_TEMPLATE_PRIORITY is not integer (ES_TEMPLATE_PRIORITY=asfsd, for instance), Spring will throw error and fail start up. Andif ES_TEMPLATE_PRIORITY= (null), then legacy template will be used.

**Comment by codefromthecrypt** [comment]
> Would you prefer ES_TEMPLATE_PRIORITY= <--- (aka null) to default to 0,
> instead of using legacy template? With the type set to nullable Integer, if
> the ES_TEMPLATE_PRIORITY is not integer (ES_TEMPLATE_PRIORITY=asfsd, for
> instance), Spring will throw error and fail start up. Andif
> ES_TEMPLATE_PRIORITY= (null), then legacy template will be used.
>
The main impact is the yml file which defaults this as then a user doesn't
need to know how it becomes null just how to make it not null :)

If spring has a way to represent a null integer, then I would recommend
using that in the template-priority ${ES_TEMPLATE_PRIORITY:XXXXXX} part for
XXXXXX
In any case a spring properties test can ensure however we are getting to a
null templatePriority works, in case spring boot changes their mind next
version about how to default things.. it will be caught.

make sense?

**Review by codefromthecrypt — APPROVED** [review]
I'm so impressed with what you've done, especially a first time committer to here. Thank you, Chanatan!

**Comment by ccharnkij** [comment]
You're welcome. Glad I can be of help.

**Comment by codefromthecrypt** [comment]
the build fail is unrelated flakey since armeria 1.0 update #3197

**Pull request merged**
